### PR TITLE
Add support for listing MT-32 model availability 

### PR DESCRIPTION
**Pull request by kcgen**
Provides a column-formatted table for which models are (or aren't) supported in those directories that have at least one available model.  Allows the users to quickly take inventory of what they have or don't have.

Because the number of models is fixed, the table shows all of them and either darkened (or lightened) depends on overall availability. 

 In the example above, we see that `mt32_106` is completely unavailable while `mt32_107` is the user's selected model.
The green highlighted 'y' additionally indicates which directory will be used during the actual loading.

### CODE REVIEW

**Comment by kcgen** [comment]
Because MT32 new and old are aliases for 204 and 107, I'm thinking we could exclude them in the table to give the directories another 8 characters of width. 

The new and old distinction is described in the conf, so I think we're not taking much away.

What do you think?

**Comment by kcgen** [comment]
For consistency with highlighting the user-selection, I will also highlight the selected romdir.

**Comment by kcgen** [comment]
> Because MT32 new and old are aliases for 204 and 107, I'm thinking we could exclude them in the table to give the directories another 8 characters of width.

Here's the result:



This adds a new array of pointers, `models_without_aliases` in the ListAll function, which the code iterates over.

**Comment by kcgen** [comment]
This is a bigger PR, so any review help - with an eye for better C++ or logical improvement - and in any capacity, would be much appreciated!

The code here is mostly business logic - getting screen dimensions, getting the roms the user has, and then printing that ascii "table" with tick-marks and colours, as shown in the screenshot above.

**Review by LowLevelMahn — COMMENTED** [review]
dont forget the munt newly added mt32 2.07 rom version

**Comment by kcgen** [comment]
> dont forget the munt newly added mt32 2.07 rom version

Oh, I'm not seeing it in the latest 2.5.0 ROM info sources: https://github.com/munt/munt/blob/munt_2_5_0/mt32emu/src/ROMInfo.cpp , but I do see it in `master` - this is great news!

When it's released in the next version, we can update the Meson wrap for it and then ask the vcpkg team to up their version too. At that point, we'll have support for it platform-wide, and then can accomodate it here too (might need an even more compact layout).

File: src/midi/midi_mt32.cpp [hunk 1]
```diff
@@ -246,13 +249,51 @@ static std::deque<std::string> get_rom_dirs()
 
 #endif
 
+static std::deque<std::string> get_selected_dirs()
+{
+	Section_prop *section = static_cast<Section_prop *>(
```
Comment (kklobe) [inline src/midi/midi_mt32.cpp:254]: Prefer `auto section = ...`?

File: src/midi/midi_mt32.cpp [hunk 2]
```diff
@@ -246,13 +249,51 @@ static std::deque<std::string> get_rom_dirs()
...
+		selected_romdir += CROSS_FILESPLIT;
+
+	// Make sure we search the user's configured directory first
+	rom_dirs.emplace_front(CROSS_ResolveHome((selected_romdir)));
+	return rom_dirs;
+}
+
+static const char *get_selected_model()
+{
+	Section_prop *section = static_cast<Section_prop *>(control->GetSection("mt32"));
```
Comment (kklobe) [inline src/midi/midi_mt32.cpp:275]: Prefer `auto section = ...`?

**Review by LowLevelMahn — APPROVED** [review]
is ok

**Review by kklobe — COMMENTED** [review]
looks solid to me.

File: src/midi/midi_mt32.cpp [hunk 3]
```diff
@@ -321,41 +362,150 @@ MidiHandler_mt32::MidiHandler_mt32()
...
+	}
+	caller->WriteOut("\n");
+
+	// Print the directories and the model(s) they support
+	num_matches = 0;
+	for (const auto &dir_and_models : dirs_with_models) {
+		const std::string &dir = dir_and_models.first;
+		const auto &dir_models = dir_and_models.second;
+
+		// Print the directory either truncated if too long
```
Comment (kcgen) [inline src/midi/midi_mt32.cpp:473]: -> `// Print the directory, and truncate it if it's too long`

File: src/midi/midi_mt32.cpp [hunk 4]
```diff
@@ -321,41 +362,150 @@ MidiHandler_mt32::MidiHandler_mt32()
...
+		if (dir.size() > max_dir_width) {
+			const auto truncated_dir = dir.substr(0, truncated_dir_width);
+			caller->WriteOut("%s%s%s", indent, truncated_dir.c_str(), trailing_dots);
+		}
+		// or in-full with padded
```
Comment (kcgen) [inline src/midi/midi_mt32.cpp:478]: -> // `Otherwise print the directory with padding`

File: src/midi/midi_mt32.cpp [hunk 5]
```diff
@@ -321,41 +362,150 @@ MidiHandler_mt32::MidiHandler_mt32()
...
+		else {
+			const auto pad_width = max_dir_width - dir.size();
+			const auto dir_padding = std::string(pad_width, ' ');
+			caller->WriteOut("%s%s%s", indent, dir.c_str(), dir_padding.c_str());
+		}
+		// Print an indicator of the directory has the model
```
Comment (kcgen) [inline src/midi/midi_mt32.cpp:484]: `of` -> `if`

**Comment by kcgen** [comment]
@LowLevelMahn , @kklobe  - thanks for the feedback.

The code isn't as simple as I would like, but I couldn't get any easier or compact.. and made several attempts.

Updates in and passing CI and tests on my side.  Merging!

**Pull request merged**